### ISSUE TITLE
Container versions may not be enough - combine with commit SHA?

### ISSUE THREAD

**Issue by rsuchecki** [csiro-crop-informatics/biokanga-manuscript#18]
Among the things we do to ensure reproducibility is to be explicit about the version of the containers we use, e.g. 

```
withLabel: rnftools {
  container = 'rsuchecki/rnftools:0.3.1.3'
}
```

There can however be scenarios where the version remains unchanged but the container has to be modified, e.g. recently `ps` became essential for some of Nextflow reporting functionality which we use for reporting run times. Most of the containers are not affected but Debian based ones are. 

We now generate additional tags for docker containers, such that in addition to `repo/tool:version` we also get `repo/tool:commitSHA`. We can now point to that, but this is not exactly readable, should we add another tag or replace that with something like `repo/tool:version_commitSHA`?

**Comment by rsuchecki** [comment]
Tools in place, need to decide if rolling out across the board.

**Comment by rsuchecki** [comment]
If using docker only, one can specify a tag along with digest, e.g. 

```
docker pull csirocropinformatics/biokanga:dev@sha256:c0475ea5827b73194852ab1d9a9e196868b852d7451cbe9f6425f8be64173bb8
```

which will return either 

* :+1:  `Status: Downloaded newer image for csirocropinformatics/biokanga@sha256:c0475ea5827b73194852ab1d9a9e196868b852d7451cbe9f6425f8be64173bb8`
* :-1:  `invalid checksum digest length`
* :-1:  `Error response from daemon: manifest for csirocropinformatics/biokanga@sha256:c0475ea5827b73194852ab1d9a9e196868b852d7451cbe9f6425f8be64173bb8 not found`

However, Singularity is not quite ready for this yet

```
FATAL:   Unable to pull docker://csirocropinformatics/biokanga:dev@sha256:c0475ea5827b73194852ab1d9a9e196868b852d7451cbe9f6425f8be64173bb8: conveyor failed to get: Invalid image sour
ce: Docker references with both a tag and digest are currently not supported
```

**Comment by rsuchecki** [comment]
A conversation clarifying some reproducibility issues when converting from Docker to Singularity and more broadly the limitations of Docker standard.

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Is building from Docker Hub reproducible if the source Docker image does not change? <a href="https://twitter.com/SingularityApp?ref_src=twsrc%5Etfw">@SingularityApp</a> <a href="https://twitter.com/vsoch?ref_src=twsrc%5Etfw">@vsoch</a> <a href="https://t.co/jDCMeVSwhE">pic.twitter.com/jDCMeVSwhE</a></p>&mdash; Rad Suchecki (@bioinforad) <a href="https://twitter.com/bioinforad/status/1120912886912536576?ref_src=twsrc%5Etfw">April 24, 2019</a></blockquote>

TLDR;
1. Singularity image you pull from a registry is a standalone binary with embedded metadata while docker image is re-assembled locally from layers bases on remote manifest. 
2. Specifying SHA sums (digests) best way to ensure consistency

**Comment by rsuchecki** [comment]
Used https://github.com/rsuchecki/miniconda3 to trial digest hash inclusion in additional tags. 

* Works and useful 
  * improved tag permanence - version+commitSHA can be accidentally overwritten from DH interface. 
  * access to digest which is otherwise not easy to get (pull or API with authorisation) 
* Could format the tag more explicitly, e.g. `1.4_sha256_050661b0ef...`, less cryptic and naturally converts to `1.4@sha256:050661b0ef...` which could be used in pulls (such a combination of tag+digest should exist).  
* Decide whether to keep commit hash alongside digest hash
